project_euler/problem_107/sol1.c

```diff
@@ -0,0 +1,188 @@
+/**
+ * \file
+ * \brief Project Euler problem 107 (https://projecteuler.net/problem=107)
+ * solution using Kruskals
+ * \author [David Santamaria](https://github.com/Daves1245)
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#define NUM_VERTICES 40
+#define NUM_LINES 40
+#define MAX_LINE_LEN 4 * 40 // a line is at most 40 3-digit numbers and a comma
+
+/*
+ * We represent an edge as a 3 tuple (u, v, w),
+ * where u and v are the vetices connected by
+ * the edge, and w is the weight of the edge
+ */
+struct edge {
+  int u;
+  int v;
+  int w;
+};
+
+/*
+ * The union find data structure
+ */
+struct ufds {
+  int rank[NUM_VERTICES], parent[NUM_VERTICES], size[NUM_VERTICES];
+  int num_sets;
+};
+
+/** Find the representative item of a vertex
+ * @param i the queried vertex
+ * @param the union find that i belongs within
+ * @returnss the index of the vertex that represents
+ * the set that i belongs to
+ */
+int find_set(int i, struct ufds *uf) {
+  if (uf->parent[i] != i) {
+    return uf->parent[i] = find_set(uf->parent[i], uf);
+  }
+  return i;
+}
+
+/** Find if i and j belong within the same set
+ * @param i the first vertex
+ * @param j the second vertex
+ * @param uf the union find i and j lie within
+ * @returns true if i and j are in the same set
+ * @returns false otherwise
+ */
+int in_same_set(int i, int j, struct ufds *uf) {
+  return find_set(i, uf) == find_set(j, uf);
+}
+
+/** Bring together two sets in a union find
+ * @param i the first representative item of one set
+ * @param j the second representative item of another set
+ * @param uf the union find they lie within
+ */
+void union_set(int i, int j, struct ufds *uf) {
+  if (!in_same_set(i, j, uf)) {
+    int ri = find_set(i, uf), rj = find_set(j, uf);
+    if (uf->rank[ri] < uf->rank[rj]) {
+      uf->parent[ri] = rj;
+      uf->size[rj]++;
+    } else {
+      uf->parent[rj] = ri;
+      if (uf->rank[ri] == uf->rank[rj]) {
+        uf->rank[rj]++;
+        uf->size[rj]++;
+      }
+    }
+    uf->num_sets--;
+  }
+}
+
+/*
+ * The problem specifies an adjacency matrix, which can be used with Prim's
+ * algorithm, but this implementation uses kruskal's, so we convert the input
+ * into an edge list represented by an array of edges
+ */
```

Comment:
```suggestion
 * @param ....
 */
```
missing parameter documentations